prometheus/reporter.go

```diff
@@ -0,0 +1,254 @@
+package prometheus
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"net/http"
+	"sort"
+	"sync"
+	"time"
+
+	prom "github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errorAlreadyRegistered = fmt.Errorf("Metric already registered")
+)
+
+type metricID string
+
+// Reporter is a prometheus backed tally reporter
+type Reporter struct {
+	counters  map[metricID]*prom.CounterVec
+	gauges    map[metricID]*prom.GaugeVec
+	summaries map[metricID]*prom.SummaryVec
+	mtx       sync.RWMutex
+}
+
+// HTTPHandler returns the prometheus HTTP handler for serving metrics
+func (r *Reporter) HTTPHandler() http.Handler {
+	return promhttp.Handler()
+}
+
+// NewReporter returns a new Reporter for Prometheus client backed metrics
+func NewReporter() Reporter {
+	counters := map[metricID]*prom.CounterVec{}
+	gauges := map[metricID]*prom.GaugeVec{}
+	summaries := map[metricID]*prom.SummaryVec{}
+	reporter := Reporter{
+		counters:  counters,
+		gauges:    gauges,
+		summaries: summaries,
+		mtx:       sync.RWMutex{},
+	}
+
+	return reporter
+}
+
+// RegisterCounter is a helper method to initialize a counter in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+func (r *Reporter) RegisterCounter(name string, tags map[string]string, desc string) (*prom.CounterVec, error) {
+	ctr := &prom.CounterVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasCounter(id)
+	if exists {
+		return ctr, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+	ctr = prom.NewCounterVec(
+		prom.CounterOpts{
+			Name: name,
+			Help: desc,
+		},
+		labelKeys,
+	)
+	err := prom.Register(ctr)
+	if err != nil {
+		return ctr, err
+	}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.counters[id] = ctr
```

Comment:
if i acquire at the top of the method, ill deadlock with the call to hasCounter. Thoughts?